app/uk/gov/hmrc/tariffclassificationfrontend/controllers/AttachmentsController.scala

```diff
@@ -0,0 +1,132 @@
+  }
+
+  private def renderView(c: Case, pForm: Form[_])
+                        (implicit hc: HeaderCarrier, request: Request[_]): Future[Html] = {
+
+    for {
+      attachments <- fileService.getAttachments(c)
+      letter <- fileService.getLetterOfAuthority(c)
+    } yield {
+      val (applicantFiles, nonApplicantFiles) = attachments.partition(_.operator.isEmpty)
+      views.html.partials.attachments_details(c, pForm, applicantFiles, letter, nonApplicantFiles)
+    }
+
+  }
+
+  private def renderErrors(reference: String, errorMessage: Option[String])
+                          (implicit hc: HeaderCarrier, req: Request[_]): Future[Result] = {
+    getCaseAndRenderView(
+      reference,
+      CaseDetailPage.ATTACHMENTS,
+      c => {
+        val errors = errorMessage match {
+          case Some(s) => Seq(FormError("file-input", s))
+          case _ => Seq.empty
+        }
+        val formWithErrors = form.copy(errors = errors)
+        renderView(c, formWithErrors)
+      }
+    )
+  }
+
+  private def uploadAndSave(reference: String, file: MultipartFormData.FilePart[TemporaryFile])
+                           (implicit hc: HeaderCarrier, request: AuthenticatedRequest[_]) = {
+
+    casesService.getOne(reference).flatMap {
+      case Some(c: Case) =>
+        fileService.upload(file).flatMap {
+          case fileStored: FileStoreAttachment =>
+            val attachments = c.attachments :+ Attachment(id = fileStored.id, operator = Some(request.operator))
+            val caseToUpdate = c.copy(attachments = attachments)
+            casesService.updateCase(caseToUpdate)
+              .flatMap( _ => successful(Redirect(routes.AttachmentsController.attachmentsDetails(reference))))
+          case _ => renderErrors(reference, Some("file service has failed while uploading."))
+        }
+      case _ => successful(Ok(views.html.case_not_found(reference)))
+    }
+  }
+
+  def uploadAttachment(reference: String): Action[MultipartFormData[TemporaryFile]] =
+    authenticatedAction.async(parse.multipartFormData) { implicit request =>
+
+      val attachment = request.body.file("file-input").filter(_.filename.nonEmpty)
+
+      attachment match {
+        case Some(file) =>
+          valid(file) match {
+            case Right(validFile) => uploadAndSave(reference, validFile)
+            case Left(errorMessage) => renderErrors(reference, Some(errorMessage))
+          }
+        case _ => renderErrors(reference, Some("no file provided."))
+      }
+    }
+
+  private val maxFileSize = 10485760
+
+  private def valid(file: MultipartFormData.FilePart[TemporaryFile]): Either[String, MultipartFormData.FilePart[TemporaryFile]] = {
+    if (file.ref.file.length() > maxFileSize) {
+      Left(messagesApi("cases.attachment.upload.restrictionSize"))
+    } else {
+      Right(file)
+    }
+  }
+
+
+  private def getCaseAndRenderView(reference: String, page: CaseDetailPage, toHtml: Case => Future[Html])
```

Comment:
Don't we have this in a super class somewhere? `RenderCaseActions` or something?

Reply:
we could potentially use a new one, because we have introduce the TAB concept